src/audio/module_adapter/module_adapter.c

```diff
@@ -169,183 +169,63 @@ static int module_adapter_sink_src_prepare(struct comp_dev *dev)
 {
 	struct comp_buffer __sparse_cache *source_buffers_c[PLATFORM_MAX_STREAMS];
 	struct comp_buffer __sparse_cache *sinks_buffers_c[PLATFORM_MAX_STREAMS];
-	struct sof_sink __sparse_cache *audio_sink[PLATFORM_MAX_STREAMS];
-	struct sof_source __sparse_cache *audio_src[PLATFORM_MAX_STREAMS];
 	struct processing_module *mod = comp_get_drvdata(dev);
 	struct list_item *blist;
-	uint32_t num_of_sources = 0;
-	uint32_t num_of_sinks = 0;
+	int i;
 	int ret;
 
 	/* acquire all sink and source buffers, get handlers to sink/source API */
+	i = 0;
 	list_for_item(blist, &dev->bsink_list) {
 		struct comp_buffer *sink_buffer_uc;
 
 		sink_buffer_uc = container_of(blist, struct comp_buffer, source_list);
-		sinks_buffers_c[num_of_sinks] = buffer_acquire(sink_buffer_uc);
-		audio_sink[num_of_sinks] =
-				audio_stream_get_sink(&sinks_buffers_c[num_of_sinks]->stream);
-		sink_reset_num_of_processed_bytes(audio_sink[num_of_sinks]);
-		num_of_sinks++;
+		sinks_buffers_c[i] = buffer_acquire(sink_buffer_uc);
+		mod->sinks[i] = audio_stream_get_sink(&sinks_buffers_c[i]->stream);
+		i++;
 	}
+	mod->num_of_sinks = i;
 
+	i = 0;
 	list_for_item(blist, &dev->bsource_list) {
 		struct comp_buffer *source_buffer_uc;
 
 		source_buffer_uc = container_of(blist, struct comp_buffer, sink_list);
-		source_buffers_c[num_of_sources] = buffer_acquire(source_buffer_uc);
-		audio_src[num_of_sources] =
-				audio_stream_get_source(&source_buffers_c[num_of_sources]->stream);
-		source_reset_num_of_processed_bytes(audio_src[num_of_sources]);
-		num_of_sources++;
+		source_buffers_c[i] = buffer_acquire(source_buffer_uc);
+		mod->sources[i] = audio_stream_get_source(&source_buffers_c[i]->stream);
```

Comment:
what happens if a modules are disconnected during unbind?

Reply:
in this case I can agree with this.. Will clear all mod->sources[] mod->sinks[] after use. Just using couple of cycles and may sometimes be safer
This is a limitation of audio_stream - as long as it need to be aquired/released. Alternative is to aquire/release each time th halder is used (inside of audio_stream itself). In my opition - too expensive. 

in case of DpQueue the handlers may be kept safely.

Reply:
> This is a limitation of audio_stream - as long as it need to be aquired/released. Alternative is to aquire/release each time th halder is used (inside of audio_stream itself). In my opition - too expensive.

If running on a single core it's cheap - as long as asserts are disabled it's a single pointer conversion when acquiring

> in case of DpQueue the handlers may be kept safely.

Sorry, kept for how long? Not over multiple `.copy()` runs? And in fact DP must be preemptive, so actually they shouldn't be kept for too long anyway